python/lsst/daf/butler/core/configSupport.py

```diff
@@ -0,0 +1,191 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Support for configuration snippets"""
+
+__all__ = ("LookupKey", "processLookupConfigs")
+
+import logging
+import re
+from collections.abc import Mapping
+from .dimensions import DimensionNameSet
+
+log = logging.getLogger(__name__)
+
+
+class LookupKey:
+    """Representation of key that can be used to lookup information based
+    on dataset type name, storage class name, dimensions.
+
+    Parameters
+    ----------
+    name : `str`, optional
+        Primary index string for lookup.  If this string looks like it
+        represents dimensions (via ``dim1+dim2+dim3`` syntax) the name
+        is converted to a `DimensionNameSet` and stored in ``dimensions``
+        property.
+    dimensions : `DimensionNameSet`, optional
+        Dimensions that are relevant for lookup. Should not be specified
+        if ``name`` is also specified.
+    dataId : `dict`, optional
+        Keys and values from a dataId that should control lookups.
+    """
+
+    def __init__(self, name=None, dimensions=None, dataId=None):
+        if name is None and dimensions is None:
+            raise ValueError("At least one of name or dimensions must be given")
+
+        if name is not None and dimensions is not None:
+            raise ValueError("Can only accept one of name or dimensions")
+
+        self._dimensions = None
+        self._name = None
+
+        if name is not None:
+            if "+" in name:
+                self._dimensions = DimensionNameSet(name.split("+"))
+            else:
+                self._name = name
+        else:
+            self._dimensions = dimensions
+
+        # The dataId is converted to a frozenset of key/value
+        # tuples.
+        if dataId is not None:
+            self._dataId = frozenset((k, v) for k, v in dataId.items())
+        else:
+            self._dataId = None
+
+    def __str__(self):
+        return "({}, {})".format(self._name if self._name else self._dimensions,
+                                 ",".join(str(t) for t in self._dataId) if self._dataId else "")
+
+    def __repr__(self):
+        params = ""
+        if self.name:
+            params += f"name={self.name!r},"
+        if self.dimensions:
+            params += f"dimensions={self.dimensions!r},"
+        if self._dataId:
+            params += "dataId={" + ",".join(f"'{k}': {v!r}" for k, v in self._dataId) + "}"
+
+        return f"{self.__class__.__name__}({params})"
+
+    def __eq__(self, other):
+        if self._name == other._name and self._dimensions == other._dimensions and \
+                self._dataId == other._dataId:
+            return True
+        return False
+
+    @property
+    def name(self):
+        """Primary name string to use as lookup."""
+        return self._name
+
+    @property
+    def dimensions(self):
+        """Dimensions associated with lookup."""
```

Comment:
Should document the type here (and other properties), especially if it's something that could either be a `DimensionNameSet` or a `DimensionGraph`, depending on whether rendezvous with `Registry.dimensions` has happened yet (or at least that's how I thought this work from Slack conversations yet; haven't gotten to that part of the code yet).